### PR TITLE
Fix crash in free() when git_buf_grow() fails.

### PR DESCRIPTION
**Pull request by jeffhostetler**
Don't try to free static buffer when realloc fails in git_buf_grow.

### CODE REVIEW

**Comment by ethomson** [comment]
Nice.  Thanks for tracking this down..

If I'm reading this correctly, this is because our first allocation was
giant and failed.  It looks like this can only happen when `buf->ptr ==
initbuf` - it looks like we couldn't have had an OOM buf to begin with
(because that would have been caught when we entered the function at
https://github.com/jeffhostetler/public_libgit2/blob/jeffhostetler/buf_oom_crash/src/buffer.c#L41
)

Can I be terrible and suggest we clean up the formatting a bit?  I don't
love `if (test) do_something` all one one line to begin with, and once that
condition gets long, I think it gets sort of unwieldly.  While we're in
there, I think that we can drop the test for `buf->ptr == oom` for clarity
(since it was tested above)?  Or am I misreading this?

File: tests/buf/oom.c [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#endif
+
+void test_buf_oom__grow(void)
+{
+	git_buf buf = GIT_BUF_INIT;
+
+	git_buf_clear(&buf);
+
+	cl_assert( git_buf_grow(&buf, TOOBIG) == -1 );
+	cl_assert( git_buf_oom(&buf) );
```
Comment (ethomson) [inline tests/buf/oom.c]: also, super pedantic - we don't put whitespace here...

**Comment by jeffhostetler** [comment]
Yeah. I included the check with oom for completeness after I tracked down the problem and didn't notice that it was already guarded at the top.  I removed it.

I still have a minor nit with the code, but I'm not sure it matters. That test on line 47 for "buf->asize == 0" is one way of saying that the internal buffer is pointing to __initbuf.  Testing for it that way up there and using the buffer later feels confusion.  But then again there is also code in git_buf_free() and git_buf_detach() that do that too.  So I'm not sure I want to get into that.

This fix should keep us from crashing on a large alloc, but doesn't address the problem of why we were asked to do that.  IIRC in the original bug it was to read a whole file into memory to compute a hash.  Seems like that code should be revisited and try to read it in chunks.

**Comment by jeffhostetler** [comment]
The appveyor failure looks like a quota/plan limit rather than an actual error.

**Comment by ethomson** [comment]
Yeah, and it seems that I can't restart appveyor builds since I'm not an owner. @nulltoken would you mind hitting the refresh button?

**Comment by nulltoken** [comment]
@ethomson Done

**Comment by carlosmn** [comment]
Squashed and merged manually. Thanks.

**Pull request closed without merging**